grails-app/services/org/pih/warehouse/inventory/InventoryService.groovy

```diff
@@ -3370,4 +3370,21 @@ class InventoryService implements ApplicationContextAware {
         return availableItems
     }
 
+    /**
+     * Get list of unique inventories that have TransactionEntries for given product
+     * */
+    List<Inventory> getInventoriesWithTransactionsByProduct(Product product) {
+        List<Transaction> transactions = Transaction.createCriteria().listDistinct {
+            transactionEntries {
+                or {
+                    eq("product", product)
+                    inventoryItem {
+                        eq("product", product)
+                    }
+                }
+            }
+        }
+
+        return transactions?.collect { it.inventory }?.unique()
```

Comment:
Can we not do this as a projection inside the query? Feels like this could be very slow with older products with a long stock history.